zrml/neo-swaps/src/tests.rs

```diff
@@ -0,0 +1,1798 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
```

Comment:
General note: The number of tests is already quite high and it is easy to lose oversight, and yet the pallet will still be extended significantly. I suggest we introduce some kind of structure that helps organizing tests and offers a reasonable level of overview. One suggestion:
```
neo-swaps
├── tests
│   ├── buy
│   │   ├── buy_works
│   │   ├── buy_fails_on_incorrect_asset_count
│   │   ├── buy_fails_on_market_not_found
│   │   ├── buy_fails_on_pool_not_found
│   │   ├── buy_fails_on_numerical_limits
│   │   ├── buy_fails_on_insufficient_funds
│   │   ├── buy_fails_on_amount_out_below_min
│   │   ├── buy_fails_on_spot_price_above_max
│   ├── deploy
│   │   ├── deploy_pool_works_with_binary_markets
│   │   ├── deploy_pool_works_with_scalar_marktes
│   │   ├── deploy_pool_fails_on_spot_price_below_min
│   │   ├── deploy_pool_fails_on_spot_price_above_max
│   │   ├── deploy_pool_fails_on_insufficient_funds
│   │   ├── deploy_pool_fails_on_not_allowed
│   │   ├── deploy_pool_fails_on_invalid_trading_mechanism
│   │   ├── deploy_pool_fails_on_market_is_not_binary_or_scalar
│   │   ├── deploy_pool_fails_on_asset_count_above_max
│   │   ├── deploy_pool_fails_on_swap_fee_below_min
│   │   ├── deploy_pool_fails_on_swap_fee_above_max
│   │   ├── deploy_pool_fails_on_incorrect_vec_len
│   │   ├── deploy_pool_fails_on_market_not_found
│   │   ├── deploy_pool_fails_on_duplicate_pool
│   │   ├── deploy_pool_fails_on_liquidity_too_low
|   ├── exit
│   │   ├── exit_works
│   │   ├── exit_destroys_pool
│   │   ├── exit_fails_on_incorrect_vec_len
│   │   ├── exit_fails_on_market_not_found
│   │   ├── exit_fails_on_pool_not_found
│   │   ├── exit_fails_on_insufficient_funds
│   │   ├── exit_fails_on_amount_out_below_min
│   │   ├── exit_fails_if_not_allowed
│   │   ├── exit_fails_on_outstanding_fees
│   │   ├── exit_pool_fails_on_liquidity_too_low
│   ├── join
│   │   ├── join_works
│   │   ├── join_fails_on_incorrect_vec_len
│   │   ├── join_fails_on_market_not_found
│   │   ├── join_fails_on_pool_not_found
│   │   ├── join_fails_on_insufficient_funds
│   │   ├── join_fails_on_amount_in_above_max
│   │   ├── join_fails_if_not_allowed
│   ├── sell
│   │   ├── sell_works
│   │   ├── sell_fails_on_incorrect_asset_count
│   │   ├── sell_fails_on_market_not_found
│   │   ├── sell_fails_on_pool_not_found
│   │   ├── sell_fails_on_numerical_limits
│   │   ├── sell_fails_on_insufficient_funds
│   │   ├── sell_fails_on_amount_out_below_min
│   │   ├── sell_fails_on_spot_price_below_min
│   ├── split
│   │   ├── split_is_noop
│   ├── withdraw
│   │   ├── withdraw_fees_works
│   │   ├── withdraw_fees_fails_on_pool_not_found
```

Reply:
Other than that, I think the unit tests presented here provide reasonable coverage.

Reply:
deec29c73baa994991ef2645d46598f28a43d1ed